File: kb_platform/api/sse.py
"""SSE (Server-Sent Events) framing shared by the streaming query/chat routes,
the MCP aggregator, and tests.

Wire format per event (blank line terminates)::

    event: <name>
    data: <one JSON line>

``data`` is a single JSON object on one line (no multi-line data), which keeps
parsing trivial for the MCP proxy and the browser reader.
"""

from __future__ import annotations

import json
from collections.abc import AsyncIterator


def format_sse(event: str, data) -> str:
    """Serialize one SSE event. ``data`` is JSON-encoded (UTF-8, non-ASCII kept)."""
    return f"event: {event}\ndata: {json.dumps(data, ensure_ascii=False)}\n\n"


def parse_sse(text: str) -> list[tuple[str, dict]]:
    """Parse a full SSE text blob into ``[(event, data), ...]``."""
    events: list[tuple[str, dict]] = []
    event: str | None = None
    data_lines: list[str] = []
    for line in text.split("\n"):
        if line.startswith("event: "):
            event = line[len("event: ") :]
        elif line.startswith("data: "):
            data_lines.append(line[len("data: ") :])
        elif line == "":
            if event is not None:
                payload = json.loads("".join(data_lines)) if data_lines else {}
                events.append((event, payload))
            event = None
            data_lines = []
    return events


async def iter_sse_events(line_aiter: AsyncIterator[str]) -> AsyncIterator[tuple[str, dict]]:
    """Parse SSE from an async line iterator (e.g. httpx ``aiter_lines()``).

    Handles a final event whose terminating blank line may be missing (the last
    chunk of a stream).
    """
    event: str | None = None
    data_lines: list[str] = []
    async for line in line_aiter:
        if line.startswith("event: "):
            # a new event implicitly terminates the previous one if no blank
            # line separated them (missing terminator on an inner event)
            if event is not None:
                payload = json.loads("".join(data_lines)) if data_lines else {}
                yield event, payload
                data_lines = []
            event = line[len("event: ") :]
        elif line.startswith("data: "):
            data_lines.append(line[len("data: ") :])
        elif line == "":
            if event is not None:
                payload = json.loads("".join(data_lines)) if data_lines else {}
                yield event, payload
            event = None
            data_lines = []
    # flush a trailing event without a terminating blank line
    if event is not None:
        payload = json.loads("".join(data_lines)) if data_lines else {}
        yield event, payload

File: web/src/lib/sse.ts
/** Parse a `text/event-stream` fetch Response into SSE events.
 *
 * `EventSource` only supports GET; our endpoints are POST with a JSON body, so
 * we read `response.body` and frame `event:` / `data:` / blank-line ourselves.
 */
export type SseEvent = { event: string; data: any };

export async function* parseSse(resp: Response): AsyncGenerator<SseEvent> {
  if (!resp.body) return;
  const reader = resp.body.getReader();
  const decoder = new TextDecoder();
  let buffer = "";
  let event = "";
  let dataLines: string[] = [];
  const flush = function* (): Generator<SseEvent> {
    if (event) {
      yield { event, data: dataLines.length ? JSON.parse(dataLines.join("")) : {} };
    }
    event = "";
    dataLines = [];
  };
  while (true) {
    const { value, done } = await reader.read();
    if (done) break;
    buffer += decoder.decode(value, { stream: true });
    const lines = buffer.split("\n");
    buffer = lines.pop() ?? ""; // keep the last (possibly partial) line
    for (const line of lines) {
      if (line.startsWith("event: ")) event = line.slice(7);
      else if (line.startsWith("data: ")) dataLines.push(line.slice(6));
      else if (line === "") yield* flush();
    }
  }
  // flush a trailing event without a terminating blank line
  buffer += decoder.decode();
  if (buffer.includes("\n")) {
    for (const line of buffer.split("\n")) {
      if (line.startsWith("event: ")) event = line.slice(7);
      else if (line.startsWith("data: ")) dataLines.push(line.slice(6));
    }
  }
  yield* flush();
}
